samplePDF/samplePDFFDBase.cpp

```diff
@@ -154,63 +144,97 @@ bool samplePDFFDBase::IsEventSelected(std::vector< std::string > ParameterStr, s
   return true;
 }
 
+#if defined (USE_PROB3) && defined (CPU_ONLY)
+
 void samplePDFFDBase::reweight(double *oscpar) // Reweight function (this should be different depending on whether you use one or 2 sets of oscpars)
 {
 
   for (int i=0; i< (int)MCSamples.size(); ++i) {
-	for(int j = 0; j < MCSamples[i].nEvents; ++j) {
-	  MCSamples[i].osc_w[j] = calcOscWeights(MCSamples[i].nutype, MCSamples[i].oscnutype, *(MCSamples[i].rw_etru[j]), oscpar);
-	}
+    for(int j = 0; j < MCSamples[i].nEvents; ++j) {
+      MCSamples[i].osc_w[j] = calcOscWeights(i, MCSamples[i].nutype, MCSamples[i].oscnutype, *(MCSamples[i].rw_etru[j]), oscpar);
+    }
   }
 
   //KS: Reset the histograms before reweight 
   ResetHistograms();
-
   fillArray();
 
   return;
 }
 
 
-void samplePDFFDBase::reweight(double *oscpar_nub, double *oscpar_nu) // Reweight function (this should be different for one vs 2 sets of oscpars)
+double samplePDFFDBase::calcOscWeights(int sample, int nutype, int oscnutype, double en, double *oscpar)
 {
-  for(int i = 0; i < (int)MCSamples.size(); ++i) {
-    for(int j = 0; j < MCSamples[i].nEvents; ++j) {
+  MCSamples[sample].Oscillator->SetMNS(oscpar[0], oscpar[2], oscpar[1], oscpar[3], oscpar[4], oscpar[5], en, doubled_angle, nutype);
+  MCSamples[sample].Oscillator->propagateLinear(nutype , oscpar[7], oscpar[8]); 
 
-		MCSamples[i].osc_w[j] = calcOscWeights(MCSamples[i].nutype, MCSamples[i].oscnutype, *(MCSamples[i].rw_etru[j]), oscpar_nub, oscpar_nu);
-    }
+  return MCSamples[sample].Oscillator->GetProb(nutype, oscnutype);
+}
+
+#endif
+
+#if defined (USE_PROB3) && not defined (CPU_ONLY)
+
+void samplePDFFDBase::reweight(double *oscpar) // Reweight function (this should be different depending on whether you use one or 2 sets of oscpars)
```

Comment:
Can we change this a bit? I find it a bit tricky to follow here. There's only ever one function for reweight, so I think it would be easier to have one definition of it but with the ifdefs around which calcOscWeight implementation should be called.

Reply:
Changed now so that ifdefs live within reweight functions